server/protocols/jmap/src/test/java/org/apache/james/jmap/utils/FilterToSearchQueryTest.java

```diff
@@ -0,0 +1,406 @@
+                            .subject(subject)
+                            .build()))
+                .build();
+
+        SearchQuery searchQuery = new FilterToSearchQuery().map(complexFilter);
+
+        assertThat(searchQuery.getCriterias()).hasSize(3);
+        assertThat(searchQuery.getCriterias().get(0)).isInstanceOf(ConjunctionCriterion.class);
+        // From
+        ConjunctionCriterion fromConjuction = (ConjunctionCriterion) searchQuery.getCriterias().get(0);
+        assertThat(fromConjuction.getCriteria()).hasSize(1);
+        assertThat(fromConjuction.getType()).isEqualTo(Conjunction.AND);
+        HeaderCriterion fromCriterion = (HeaderCriterion) fromConjuction.getCriteria().get(0);
+        assertThat(fromCriterion.getHeaderName()).isEqualTo(AddressType.From.name());
+        assertThat(fromCriterion.getOperator()).isInstanceOf(AddressOperator.class).isEqualTo(new AddressOperator(from));
+        // To
+        ConjunctionCriterion toConjuction = (ConjunctionCriterion) searchQuery.getCriterias().get(1);
+        assertThat(toConjuction.getCriteria()).hasSize(1);
+        assertThat(toConjuction.getType()).isEqualTo(Conjunction.AND);
+        HeaderCriterion toCriterion = (HeaderCriterion) toConjuction.getCriteria().get(0);
+        assertThat(toCriterion.getHeaderName()).isEqualTo(AddressType.To.name());
+        assertThat(toCriterion.getOperator()).isInstanceOf(AddressOperator.class).isEqualTo(new AddressOperator(to));
+        // Subject
+        ConjunctionCriterion subjectConjuction = (ConjunctionCriterion) searchQuery.getCriterias().get(2);
+        assertThat(subjectConjuction.getCriteria()).hasSize(1);
+        assertThat(subjectConjuction.getType()).isEqualTo(Conjunction.AND);
+        HeaderCriterion subjectCriterion = (HeaderCriterion) subjectConjuction.getCriteria().get(0);
+        assertThat(subjectCriterion.getHeaderName()).isEqualTo("Subject");
+        assertThat(subjectCriterion.getOperator()).isInstanceOf(ContainsOperator.class).isEqualTo(new ContainsOperator(subject));
+    }
+
+    @Test
+    public void filterConditionShouldMapWhenOrOperator() {
+        String from = "sender@james.org";
+        String to = "recipient@james.org";
+        String subject = "subject";
+        Filter complexFilter = FilterOperator.builder()
+                .operator(Operator.OR)
+                .conditions(ImmutableList.of(
+                        FilterCondition.builder()
+                            .from(from)
+                            .build(),
+                        FilterCondition.builder()
+                            .to(to)
+                            .build(),
+                        FilterCondition.builder()
+                            .subject(subject)
+                            .build()))
+                .build();
+
+        SearchQuery searchQuery = new FilterToSearchQuery().map(complexFilter);
+
+        assertThat(searchQuery.getCriterias()).hasSize(3);
```

Comment:
You definitely have to build a query tree and do a equals instead of checking every single part